File: src/Algorithm/LeetCode/LeetCode_739_Daily_Temperatures/Solution.java
package Algorithm.LeetCode.LeetCode_739_Daily_Temperatures;

import java.util.Stack;

// check LeetCode_503
class Solution {
    public int[] dailyTemperatures(int[] temperatures) {
        int n = temperatures.length;
        int[] result = new int[n];

        if (n == 1) {
            return result;
        }

        Stack<Integer> stack = new Stack<>();

        for (int i = 0; i < n; i++) {
            while (!stack.isEmpty() && temperatures[stack.peek()] < temperatures[i]) {
                int curr = stack.pop();
                result[curr] = i - curr;
            }

            stack.push(i);
        }

        return result;
    }
}

File: src/Algorithm/LeetCode/LeetCode_739_Daily_Temperatures/Solution.py
# https://www.programmercarl.com/0739.每日温度.html
class Solution:
    def dailyTemperatures(self, temperatures: List[int]) -> List[int]:
        length = len(temperatures)
        result = [0] * length
        stack = [0]

        for i in range(1, length):
            while len(stack) > 0 and temperatures[i] > temperatures[stack[-1]]:
                result[stack[-1]] = i - stack[-1]
                stack.pop()

            stack.append(i)

        return result
